lib/compat/wordpress-6.5/class-block-search-controller.php

```diff
@@ -0,0 +1,171 @@
+<?php
+/**
+ * Block search controller.
+ *
+ * @package Block_Search_Controller
+ */
+
+if ( ! class_exists( 'Block_Search_Controller' ) ) {
+	/**
+	 * Class Block_Search_Controller.
+	 */
+	class Block_Search_Controller extends WP_REST_Controller {
+
+		/**
+		 * Constructor.
+		 */
+		public function __construct() {
+			$this->namespace = 'wp/v2';
+			$this->rest_base = 'block_search';
+		}
+
+		/**
+		 * Registers the routes for the objects of the controller.
+		 */
+		public function register_routes() {
+			register_rest_route(
+				$this->namespace,
+				'/' . $this->rest_base,
+				array(
+					array(
+						'methods'             => WP_REST_Server::READABLE,
+						'callback'            => array(
+							$this,
+							'find_posts_by_block_attribute',
+						),
+						'args'                => array(
+							'post_type'       => array(
+								'required' => true,
+							),
+							'block_name'      => array(
+								'required' => true,
+							),
+							'attribute_name'  => array(
+								'required' => true,
+							),
+							'attribute_value' => array(
+								'required' => true,
+							),
+							'posts_per_page'  => array(
+								'required' => false,
+							),
+						),
+						'permission_callback' => array( $this, 'permissions_check' ),
+					),
+				)
+			);
+		}
+
+		/**
+		 * Permission callback to check if the user can use this endpoint.
+		 */
+		public function permissions_check() {
+			// Check user permissions.
+			return true;
+		}
+
+		/**
+		 * The callback for the find_posts_by_block route.
+		 *
+		 * @param WP_REST_Request $request The request object.
+		 */
+		public function find_posts_by_block_attribute( $request ) {
+			$post_type       = $request['post_type'];
+			$block_name      = $request['block_name'];
+			$attribute_name  = $request['attribute_name'];
+			$attribute_value = $request['attribute_value'];
+
+			$posts_per_page = apply_filters(
+				'block_search_posts_per_page',
+				$request['posts_per_page'] ?? get_option( 'posts_per_page' ),
+				$request
+			);
+
+			$paged            = 1; // Start on the first page.
+			$posts_with_block = array();
+
+			do {
+				$posts = get_posts(
+					array(
+						'post_type'      => $post_type,
+						'posts_per_page' => $posts_per_page,
+						'paged'          => $paged,
+						'post_status'    => 'publish',
+					)
+				);
+
+				if ( empty( $posts ) ) {
+					break;
+				}
+
+				foreach ( $posts as $post ) {
+					if ( count( $posts_with_block ) >= $posts_per_page ) {
+						break 2;
+					}
+
+					$blocks       = parse_blocks( $post->post_content );
+					$found_blocks = $this->recursively_find_block_by_attribute(
+						$blocks,
+						$block_name,
+						$attribute_name,
+						$attribute_value
+					);
```

Comment:
@draganescu I've been exploring a lazy parser with significantly better performance characteristics, especially when only wanting to find blocks in a post. it needs work before it can be proposed, but it's great because it does what I proposed here but is memory efficient and can stop once we've found what we're looking for.

WordPress/wordpress-develop#5705

(I do plan on porting this soon into the Gutenberg repository)